language/solana/move-to-solana/src/stackless/entrypoint.rs

```diff
@@ -0,0 +1,525 @@
+// Copyright (c) The Diem Core Contributors
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    options::Options,
+    stackless::{
+        extensions::*, llvm, module_context::ModuleContext, rttydesc::RttyContext, GlobalContext,
+        RtCall,
+    },
+};
+use log::{debug, log_enabled, Level};
+use move_core_types::u256::U256;
+use move_model::{model as mm, ty as mty};
+use move_native::shared::MOVE_UNTYPED_VEC_DESC_SIZE;
+use std::{
+    cell::{Cell, RefCell},
+    collections::BTreeMap,
+    path::Path,
+};
+
+pub struct EntrypointGenerator<'mm, 'up> {
+    pub env: &'mm mm::GlobalEnv,
+    pub llvm_cx: &'up llvm::Context,
+    pub llvm_module: &'up llvm::Module,
+    pub llvm_builder: llvm::Builder,
+    pub options: &'up Options,
+    pub rtty_cx: RttyContext<'mm, 'up>,
+
+    fn_decls: RefCell<BTreeMap<String, llvm::Function>>,
+
+    ll_fn_solana_entrypoint: llvm::Function,
+    entry_slice_ptr: llvm::AnyValue,
+    entry_slice_len: llvm::AnyValue,
+    insn_data_ptr: llvm::AnyValue,
+    offset: llvm::Alloca,
+    retval: llvm::AnyValue,
+    exit_bb: llvm::BasicBlock,
+
+    entries: Cell<u64>,
+    target_machine: &'up llvm::TargetMachine,
+}
+
+/**
+ * Generate solana entrypoint functon code. This function
+ * recieves serialized input paramteres from the VM. It calls
+ * native function `deserialize` to decode the parameters into
+ * corresponding data structures. The function `deserialize`
+ * returns a triple consiting of
+ * - instruction_data -- a byte array,
+ * - program_id -- SolanaPubkey, and
+ * - accounts -- a vector of SolanaAccountInfo items.
+ *
+ * To select one from possibly several entry functions defined in
+ * the module, the entrypoint function expects the name of the
+ * requested entry function to be passed in instruction_data byte
+ * array. The logic in solana entrypoint iteratively compares the
+ * string slice passed in instruction_data to every entry function
+ * symbol of the module. Once a matching entry function is found,
```

Comment:
if two modules have the same name and they both contain entry functions with the same name, then yes there will be duplicates. We probably will include modules’ unique addresses in function names to avoid conflicts.

Reply:
I added a `debug_assert!()` to check for duplicate entry function names.